File: frontend/src/api/repo.ts
import { apiRequest } from './client'

export type RepoTreeEntry = {
  path: string
  name: string
  is_dir: boolean
  size: number | null
}

export type RepoTreeResponse = {
  root: string
  entries: RepoTreeEntry[]
}

export type RepoFileContent = {
  path: string
  content: string
  size: number
  truncated: boolean
}

export function getRepoTree(projectId: string, maxDepth?: number) {
  const query =
    maxDepth === undefined ? '' : `?max_depth=${encodeURIComponent(String(maxDepth))}`
  return apiRequest<RepoTreeResponse>(`/api/v1/projects/${projectId}/repo/tree${query}`)
}

export function getRepoFile(projectId: string, path: string) {
  return apiRequest<RepoFileContent>(
    `/api/v1/projects/${projectId}/repo/file?path=${encodeURIComponent(path)}`,
  )
}

File: backend/app/schemas/repo.py
"""Repository browsing schemas."""

from pydantic import BaseModel, Field


class RepoTreeEntry(BaseModel):
    path: str
    name: str
    is_dir: bool
    size: int | None = None


class RepoTreeResponse(BaseModel):
    root: str
    entries: list[RepoTreeEntry] = Field(default_factory=list)


class RepoFileContent(BaseModel):
    path: str
    content: str
    size: int
    truncated: bool = False
